azure/services/agentpools/agentpools.go

```diff
@@ -80,8 +81,9 @@ func (s *Service) Reconcile(ctx context.Context, spec interface{}) error {
 	} else {
 		ps := *existingPool.ManagedClusterAgentPoolProfileProperties.ProvisioningState
 		if ps != "Canceled" && ps != "Failed" && ps != "Succeeded" {
-			klog.V(2).Infof("Unable to update existing agent pool in non terminal state.  Agent pool must be in one of the following provisioning states: canceled, failed, or succeeded")
-			return nil
+			msg := fmt.Sprintf("Unable to update existing agent pool in non terminal state. Agent pool must be in one of the following provisioning states: canceled, failed, or succeeded. Actual state: %s", ps)
+			klog.V(2).Infof(msg)
+			return errors.New(msg)
```

Comment:
Perhaps, this error should return an transient error wrapper to inform the controller to requeue after a period of time. This would cause an immediate requeue, right?

Reply:
immediate requeue isn't great, i'm kind of fine with the exponential backoff though tbh. I'm fine with doing an error wrapper here for the short term, if you'd like.

but...I really don't love generating errors in low level clients which higher level controller code needs to know about. I'd almost prefer to return reconcile.Results{} all the way up/down so the lower level clients can provide useful semantics to the controllers. It's annoying if we add new errors and have to check for them in the top level, we've already had bugs in because we didn't check for `ResourceGroupNotFound/ResourceNotFound` correctly at the top level after we added the change in the lower level (which you kindly fixed) https://github.com/kubernetes-sigs/cluster-api-provider-azure/commit/f1e81c1ac979ba57c24642ce9fd8430449817e58#diff-714daa93f2eede09b5665786a53bcddf09a13a8d0a19cd17fc61e3bc017f4b8aR198

Reply:
fwiw, I didn't notice much negative effect in #1488 -- the timing work out pretty closely to how long AKS take to create the cluster.

Reply:
well, I guess this is only hit on update/delete anyway, so maybe that's not meaningful.

Reply:
are we already handling terminal/transient errors in the managed cluster reconciler?

Reply:
https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/e1faf08f76253e8b4200340a41b2e26dd86867f5/exp/controllers/azuremangedcontrolplane_reconciler.go#L207-L215

only to that extent in AMCP

https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/e1faf08f76253e8b4200340a41b2e26dd86867f5/exp/controllers/azuremanagedmachinepool_reconciler.go#L145

and 404s on VMSS. 

we aren't doing anything on provisioning state at the higher levels, which is why this is a problem(/bug), because there's no signal for a requeue when provisioning state has reached a terminal state.

Reply:
this is what I meant (and I assume what David was also referring to) by transient/terminal error handling: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/controllers/azuremachine_controller.go#L284

I don't think these are being used anywhere outside of azuremachine and azuremachinepool controllers though (even azurecluster controller doesn't have logic to handle them) so I'm good with the bug fix as-is and we can revisit introducing terminal/transient error handling in the managedcluster controllers separately. I do think it would be beneficial as @devigned pointed out to be able to not requeue immediately on these types of errors where a bit of waiting is necessary.